File: Loenn/entities/SizableJelly.lua
local drawableSprite = require("structs.drawable_sprite")
local drawableLine = require("structs.drawable_line")
local drawing = require("utils.drawing")
local utils = require('utils')

local glider = {}

glider.name = "VivHelper/SizableJelly"
glider.depth = -5
glider.placements = {
    {
        name = "normal",
        data = {
            tutorial = false,
            bubble = false,
            width = 8, height = 8,
            scaleGrabBoxWithHitbox = true
        }
    },
    {
        name = "floating",
        data = {
            tutorial = false,
            bubble = true,
            width = 8, height = 8,
            scaleGrabBoxWithHitbox = true
        }
    }
}

local texture = "objects/glider/idle0"

function glider.sprite(room, entity)
    local bubble = entity.bubble

    if entity.bubble then
        local x, y = entity.x or 0, entity.y or 0
        local points = drawing.getSimpleCurve({x - 11, y - 1}, {x + 11, y - 1}, {x - 0, y - 6})
        local lineSprites = drawableLine.fromPoints(points):getDrawableSprite()
        local jellySprite = drawableSprite.fromTexture(texture, entity)
        jellySprite:setScale(entity.width/8,entity.height/8)

        table.insert(lineSprites, 1, jellySprite)

        return lineSprites

    else
        local jellySprite = drawableSprite.fromTexture(texture, entity)
        jellySprite:setScale(entity.width/8,entity.height/8)
        return jellySprite
    end
end

function glider.selection(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local sx, sy = entity.width or 8, entity.height or 8
    return utils.rectangle(x - sx/2, y - sy, sx, sy)
end

return glider

File: Ahorn/entities/SizableJelly.jl
module VH_SizableJellyfish

using ..Ahorn, Maple

@mapdef Entity "VivHelper/SizableJelly" SizableJelly(x::Integer, y::Integer, width::Integer=16, height::Integer=16, bubble::Bool=false, tutorial::Bool=false, scaleGrabBoxWithHitbox::Bool=true)

const placements = Ahorn.PlacementDict(
    "Sizable Jellyfish" => Ahorn.EntityPlacement(
        SizableJelly
    ),
    "Sizable Jellyfish (Floating)" => Ahorn.EntityPlacement(
        SizableJelly,
        "point",
        Dict{String, Any}(
            "bubble" => true
        )
    )
)

Ahorn.resizable(entity::SizableJelly)

sprite = "objects/glider/idle0"

function Ahorn.selection(entity::SizableJelly)
    x, y = Ahorn.position(entity)
    return Ahorn.Rectangle(x,y, get(entity, "width", 8), get(entity, "height", 8))
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::SizableJelly, room::Maple.Room)
    x, y = Ahorn.position(entity)
    w, h = get(entity, "width", 8),get(entity, "height", 8)
    Ahorn.drawSprite(ctx, sprite, x + w/2, y+h/2; jx=0.5, jy=0.5, sx=w/8, sy=h/8)

    if get(entity, "bubble", false)
        curve = Ahorn.SimpleCurve((-7, -1), (7, -1), (0, -6))
        Ahorn.drawSimpleCurve(ctx, curve, (1.0, 1.0, 1.0, 1.0), thickness=1)
    end
end

end